File: controllers/ftp_controller.py
from flask import Blueprint, render_template, request, redirect, url_for, flash, jsonify
from flask_login import login_required, current_user
from werkzeug.utils import secure_filename
import os
from services import FTPService, TransferService
from functools import wraps

ftp_bp = Blueprint('ftp', __name__)

def validate_file_type(filename):
    allowed_extensions = {'txt', 'pdf', 'png', 'jpg', 'jpeg', 'gif', 'doc', 'docx', 'xls', 'xlsx', 'zip', 'rar', 'mp4', 'avi', 'mov', 'mp3', 'wav', 'csv', 'json', 'xml', 'html', 'css', 'js', 'py', 'java', 'cpp', 'c', 'sql', 'log'}
    if '.' not in filename:
        return False
    extension = filename.rsplit('.', 1)[1].lower()
    return extension in allowed_extensions

@ftp_bp.route('/api/upload/<int:account_id>', methods=['POST'])
@login_required
def upload_file(account_id):
    account = FTPService.get_account_by_id(account_id)
    if not account:
        return jsonify({'success': False, 'message': 'FTP hesabı bulunamadı'})
    
    # Check if user has access to this account
    assigned_users = FTPService.get_account_assignments(account_id)
    if current_user not in assigned_users:
        return jsonify({'success': False, 'message': 'Bu hesaba erişim yetkiniz yok'})
    
    if 'file' not in request.files:
        return jsonify({'success': False, 'message': 'Dosya seçilmedi'})
    
    file = request.files['file']
    if file.filename == '':
        return jsonify({'success': False, 'message': 'Dosya seçilmedi'})
    
    try:
        filename = secure_filename(file.filename)
        
        # Validate file type
        if not validate_file_type(filename):
            return jsonify({'success': False, 'message': 'Desteklenmeyen dosya türü'})
        
        file_path = os.path.join('uploads', filename)
        file.save(file_path)
        
        # Upload to FTP/SFTP server
        if account.protocol == 'ftp':
            success, error = FTPService.upload_to_ftp(account, file_path, filename)
        elif account.protocol == 'sftp':
            success, error = FTPService.upload_to_sftp(account, file_path, filename)
        else:
            return jsonify({'success': False, 'message': 'Desteklenmeyen protokol'})
        
        if success:
            # Log the transfer
            TransferService.log_transfer(current_user.id, account_id, 'upload', filename, os.path.getsize(file_path), 'success')
            
            # Clean up local file
            os.remove(file_path)
            
            return jsonify({'success': True, 'message': 'Dosya başarıyla yüklendi'})
        else:
            # Clean up local file if exists
            if os.path.exists(file_path):
                os.remove(file_path)
            
            TransferService.log_transfer(current_user.id, account_id, 'upload', filename, 0, 'error', error)
            return jsonify({'success': False, 'message': error})
    except Exception as e:
        # Clean up local file if exists
        if 'file_path' in locals() and os.path.exists(file_path):
            os.remove(file_path)
        
        TransferService.log_transfer(current_user.id, account_id, 'upload', filename, 0, 'error', str(e))
        return jsonify({'success': False, 'message': str(e)})

@ftp_bp.route('/api/download/<int:account_id>/<filename>')
@login_required
def download_file(account_id, filename):
    account = FTPService.get_account_by_id(account_id)
    if not account:
        return jsonify({'success': False, 'message': 'FTP hesabı bulunamadı'})
    
    # Check if user has access to this account
    assigned_users = FTPService.get_account_assignments(account_id)
    if current_user not in assigned_users:
        return jsonify({'success': False, 'message': 'Bu hesaba erişim yetkiniz yok'})
    
    try:
        if account.protocol == 'ftp':
            file_data, error = FTPService.download_from_ftp(account, filename)
        elif account.protocol == 'sftp':
            file_data, error = FTPService.download_from_sftp(account, filename)
        else:
            return jsonify({'success': False, 'message': 'Desteklenmeyen protokol'})
        
        if file_data is not None:
            TransferService.log_transfer(current_user.id, account_id, 'download', filename, len(file_data), 'success')
            
            from flask import Response
            return Response(
                file_data,
                mimetype='application/octet-stream',
                headers={'Content-Disposition': f'attachment; filename={filename}'}
            )
        else:
            TransferService.log_transfer(current_user.id, account_id, 'download', filename, 0, 'error', error)
            return jsonify({'success': False, 'message': error})
    except Exception as e:
        TransferService.log_transfer(current_user.id, account_id, 'download', filename, 0, 'error', str(e))
        return jsonify({'success': False, 'message': str(e)})

@ftp_bp.route('/api/delete/<int:account_id>/<filename>', methods=['DELETE'])
@login_required
def delete_file(account_id, filename):
    # Only admin users can delete files
    if current_user.role != 'admin':
        return jsonify({'success': False, 'message': 'Dosya silme yetkiniz yok. Sadece admin kullanıcıları dosya silebilir.'})
    
    account = FTPService.get_account_by_id(account_id)
    if not account:
        return jsonify({'success': False, 'message': 'FTP hesabı bulunamadı'})
    
    # Check if admin has access to this account (admin can delete from any account they own)
    if account.owner_id != current_user.id:
        return jsonify({'success': False, 'message': 'Bu hesaptan dosya silme yetkiniz yok'})
    
    try:
        if account.protocol == 'ftp':
            success, error = FTPService.delete_from_ftp(account, filename)
        elif account.protocol == 'sftp':
            success, error = FTPService.delete_from_sftp(account, filename)
        else:
            return jsonify({'success': False, 'message': 'Desteklenmeyen protokol'})
        
        if success:
            TransferService.log_transfer(current_user.id, account_id, 'delete', filename, 0, 'success')
            return jsonify({'success': True, 'message': 'Dosya başarıyla silindi'})
        else:
            TransferService.log_transfer(current_user.id, account_id, 'delete', filename, 0, 'error', error)
            return jsonify({'success': False, 'message': error})
    except Exception as e:
        TransferService.log_transfer(current_user.id, account_id, 'delete', filename, 0, 'error', str(e))
        return jsonify({'success': False, 'message': str(e)})

File: java_version/src/main/java/com/xfer/controller/FTPController.java
package com.xfer.controller;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import com.xfer.entity.FTPAccount;
import com.xfer.entity.User;
import com.xfer.service.FTPService;
import com.xfer.service.TransferService;

@RestController
@RequestMapping("/api")
public class FTPController {
    
    @Autowired
    private FTPService ftpService;
    
    
    @Autowired
    private TransferService transferService;
    
    private static final String UPLOAD_DIR = "uploads";
    
    @PostMapping("/upload/{accountId}")
    public ResponseEntity<Map<String, Object>> uploadFile(
            @PathVariable Long accountId,
            @RequestParam("file") MultipartFile file,
            Authentication authentication) {
        
        Map<String, Object> response = new HashMap<>();
        User currentUser = (User) authentication.getPrincipal();
        
        try {
            // Check if user has access to this account
            if (!ftpService.hasUserAccess(accountId, currentUser.getId())) {
                response.put("success", false);
                response.put("message", "Bu hesaba erişim yetkiniz yok");
                return ResponseEntity.badRequest().body(response);
            }
            
            Optional<FTPAccount> accountOpt = ftpService.getAccountById(accountId);
            if (!accountOpt.isPresent()) {
                response.put("success", false);
                response.put("message", "FTP hesabı bulunamadı");
                return ResponseEntity.badRequest().body(response);
            }
            
            FTPAccount account = accountOpt.get();
            
            if (file.isEmpty()) {
                response.put("success", false);
                response.put("message", "Dosya seçilmedi");
                return ResponseEntity.badRequest().body(response);
            }
            
            // Validate file type
            if (!isValidFileType(file.getOriginalFilename())) {
                response.put("success", false);
                response.put("message", "Desteklenmeyen dosya türü");
                return ResponseEntity.badRequest().body(response);
            }
            
            // Create upload directory if it doesn't exist
            Path uploadPath = Paths.get(UPLOAD_DIR);
            if (!Files.exists(uploadPath)) {
                Files.createDirectories(uploadPath);
            }
            
            // Save file temporarily
            String filename = file.getOriginalFilename();
            Path filePath = uploadPath.resolve(filename);
            file.transferTo(filePath.toFile());
            
            try {
                // Upload to FTP/SFTP server
                boolean success = ftpService.uploadFile(accountId, file);
                
                if (success) {
                    // Log the transfer
                    transferService.logTransfer(currentUser.getId(), accountId, "upload", 
                            filename, file.getSize(), "success", null);
                    
                    response.put("success", true);
                    response.put("message", "Dosya başarıyla yüklendi");
                } else {
                    transferService.logTransfer(currentUser.getId(), accountId, "upload", 
                            filename, 0L, "error", "Yükleme başarısız");
                    
                    response.put("success", false);
                    response.put("message", "Dosya yüklenirken hata oluştu");
                }
                
            } finally {
                // Clean up local file
                Files.deleteIfExists(filePath);
            }
            
        } catch (Exception e) {
            transferService.logTransfer(currentUser.getId(), accountId, "upload", 
                    file.getOriginalFilename(), 0L, "error", e.getMessage());
            
            response.put("success", false);
            response.put("message", "Hata: " + e.getMessage());
        }
        
        return ResponseEntity.ok(response);
    }
    
    @GetMapping("/download/{accountId}/{filename}")
    public ResponseEntity<Resource> downloadFile(
            @PathVariable Long accountId,
            @PathVariable String filename,
            Authentication authentication) {
        
        User currentUser = (User) authentication.getPrincipal();
        
        try {
            // Check if user has access to this account
            if (!ftpService.hasUserAccess(accountId, currentUser.getId())) {
                return ResponseEntity.badRequest().build();
            }
            
            Optional<FTPAccount> accountOpt = ftpService.getAccountById(accountId);
            if (!accountOpt.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            FTPAccount account = accountOpt.get();
            
            byte[] fileData = ftpService.downloadFile(accountId, filename);
            
            // Log the transfer
            transferService.logTransfer(currentUser.getId(), accountId, "download", 
                    filename, (long) fileData.length, "success", null);
            
            ByteArrayResource resource = new ByteArrayResource(fileData);
            
            return ResponseEntity.ok()
                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + filename + "\"")
                    .contentType(MediaType.APPLICATION_OCTET_STREAM)
                    .contentLength(fileData.length)
                    .body(resource);
                    
        } catch (Exception e) {
            transferService.logTransfer(currentUser.getId(), accountId, "download", 
                    filename, 0L, "error", e.getMessage());
            
            return ResponseEntity.badRequest().build();
        }
    }
    
    @DeleteMapping("/delete/{accountId}/{filename}")
    public ResponseEntity<Map<String, Object>> deleteFile(
            @PathVariable Long accountId,
            @PathVariable String filename,
            Authentication authentication) {
        
        Map<String, Object> response = new HashMap<>();
        User currentUser = (User) authentication.getPrincipal();
        
        // Only admin users can delete files
        if (!"admin".equals(currentUser.getRole())) {
            response.put("success", false);
            response.put("message", "Dosya silme yetkiniz yok. Sadece admin kullanıcıları dosya silebilir.");
            return ResponseEntity.badRequest().body(response);
        }
        
        try {
            Optional<FTPAccount> accountOpt = ftpService.getAccountById(accountId);
            if (!accountOpt.isPresent()) {
                response.put("success", false);
                response.put("message", "FTP hesabı bulunamadı");
                return ResponseEntity.badRequest().body(response);
            }
            
            FTPAccount account = accountOpt.get();
            
            // Check if admin has access to this account
            if (!account.getOwnerId().equals(currentUser.getId())) {
                response.put("success", false);
                response.put("message", "Bu hesaptan dosya silme yetkiniz yok");
                return ResponseEntity.badRequest().body(response);
            }
            
            boolean success = ftpService.deleteFile(accountId, filename);
            
            if (success) {
                transferService.logTransfer(currentUser.getId(), accountId, "delete", 
                        filename, 0L, "success", null);
                
                response.put("success", true);
                response.put("message", "Dosya başarıyla silindi");
            } else {
                transferService.logTransfer(currentUser.getId(), accountId, "delete", 
                        filename, 0L, "error", "Silme başarısız");
                
                response.put("success", false);
                response.put("message", "Dosya silinirken hata oluştu");
            }
            
        } catch (Exception e) {
            transferService.logTransfer(currentUser.getId(), accountId, "delete", 
                    filename, 0L, "error", e.getMessage());
            
            response.put("success", false);
            response.put("message", "Hata: " + e.getMessage());
        }
        
        return ResponseEntity.ok(response);
    }
    
    @GetMapping("/list/{accountId}")
    public ResponseEntity<Map<String, Object>> listFiles(
            @PathVariable Long accountId,
            Authentication authentication) {
        
        Map<String, Object> response = new HashMap<>();
        User currentUser = (User) authentication.getPrincipal();
        
        try {
            // Check if user has access to this account
            if (!ftpService.hasUserAccess(accountId, currentUser.getId())) {
                response.put("success", false);
                response.put("message", "Bu hesaba erişim yetkiniz yok");
                return ResponseEntity.badRequest().body(response);
            }
            
            Optional<FTPAccount> accountOpt = ftpService.getAccountById(accountId);
            if (!accountOpt.isPresent()) {
                response.put("success", false);
                response.put("message", "FTP hesabı bulunamadı");
                return ResponseEntity.badRequest().body(response);
            }
            
            FTPAccount account = accountOpt.get();
            
            List<FTPService.FileInfo> files = ftpService.listFiles(accountId);
            
            // Log the transfer
            transferService.logTransfer(currentUser.getId(), accountId, "list", 
                    null, 0L, "success", null);
            
            response.put("success", true);
            response.put("files", files);
            
        } catch (Exception e) {
            transferService.logTransfer(currentUser.getId(), accountId, "list", 
                    null, 0L, "error", e.getMessage());
            
            response.put("success", false);
            response.put("message", "Hata: " + e.getMessage());
        }
        
        return ResponseEntity.ok(response);
    }
    
    private boolean isValidFileType(String filename) {
        if (filename == null || !filename.contains(".")) {
            return false;
        }
        
        String extension = filename.substring(filename.lastIndexOf(".") + 1).toLowerCase();
        String[] allowedExtensions = {
            "xls", "xlsx","csv", "json", "xml","sql"
        };
        
        for (String allowedExt : allowedExtensions) {
            if (allowedExt.equals(extension)) {
                return true;
            }
        }
        
        return false;
    }
}
